File: sdk/typescript/src/human.ts
/**
 * The human side of both flows without a browser: exactly what the drop
 * page does, with the same envelope and the same checks. Useful for tests,
 * scripts, and command line tools. Mirrors cmd/burndrop/cmd_human.go.
 */
import {
  commitment,
  decryptEnvelope,
  fingerprint,
  revealAad,
  sealEnvelope,
  secretBytes,
  secretField,
  zero,
  type Envelope,
  type SecretFormat,
} from "./crypto.js";
import { BurndropError, RelayError, ValidationError, isRelayCode } from "./errors.js";
import { parseDropLink, parseRevealLink, relayOrigin } from "./link.js";
import { RelayClient, RelayCode, SlotState, type RelayClientOptions } from "./relay.js";

export interface HumanOptions extends RelayClientOptions {
  /** Talk to this relay instead of the one named by the link. */
  relay?: string;
  /** Abort signal for the relay calls. */
  signal?: AbortSignal;
}

export interface SubmitResult {
  name: string;
  purpose: string;
  storage: string;
  retention: string;
  /** Fingerprint of the key the value was encrypted to; compare it with the agent's message. */
  fingerprint: string;
  sizeBytes: number;
  relay: string;
}

export interface OpenResult {
  name: string;
  value: Uint8Array;
  format: SecretFormat;
  keepsCopy: boolean;
  relay: string;
}

const utf8 = new TextEncoder();

function makeClient(origin: string, options: HumanOptions): RelayClient {
  const clientOptions: RelayClientOptions = {};
  if (options.clientName !== undefined) {
    clientOptions.clientName = options.clientName;
  }
  if (options.fetch !== undefined) {
    clientOptions.fetch = options.fetch;
  }
  if (options.timeoutMs !== undefined) {
    clientOptions.timeoutMs = options.timeoutMs;
  }
  return new RelayClient(origin, "", clientOptions);
}

/**
 * Submits a value for a drop link: checks that the link is still open,
 * builds the envelope from the link's display fields and the fingerprint
 * of its key, seals it to that key, and uploads it with the key's
 * commitment. The relay refuses the upload when the commitment differs
 * from the one the agent registered, which is how an altered link is
 * caught; that refusal surfaces here as a BurndropError.
 */
export async function submit(link: string, value: Uint8Array | string, options: HumanOptions = {}): Promise<SubmitResult> {
  const { drop, pageOrigin } = parseDropLink(link);
  const relay = options.relay ?? relayOrigin(drop, pageOrigin);
  const client = makeClient(relay, options);
  const bytes = typeof value === "string" ? utf8.encode(value) : value;
  if (bytes.length === 0) {
    throw new ValidationError("empty value");
  }
  let status;
  try {
    status = await client.dropStatus(drop.id, 0, "", options.signal);
  } catch (err) {
    if (isRelayCode(err, RelayCode.NotFound)) {
      throw new BurndropError("this link has expired or was never created", { cause: err });
    }
    throw err;
  }
  if (status.state !== SlotState.Created) {
    throw new BurndropError("this link can no longer be used (state: " + status.state + ")");
  }
  const fp = await fingerprint(drop.recipientKey);
  const env: Envelope = {
    v: 1,
    type: "drop",
    name: drop.name,
    purpose: drop.purpose,
    storage: drop.storage,
    retention: drop.retention,
    fingerprint: fp,
    ...secretField(bytes),
  };
  const sealed = await sealEnvelope(drop.recipientKey, env);
  try {
    await client.upload(drop.id, drop.uploadToken, await commitment(drop.recipientKey), sealed, options.signal);
  } catch (err) {
    if (isRelayCode(err, RelayCode.CommitmentMismatch)) {
      throw new BurndropError("the key in this link does not match the key the agent registered; the link was altered, do not use it", { cause: err });
    }
    throw err;
  }
  return {
    name: drop.name,
    purpose: drop.purpose,
    storage: drop.storage,
    retention: drop.retention,
    fingerprint: fp,
    sizeBytes: bytes.length,
    relay,
  };
}

/**
 * Opens a reveal link: checks that it is still unopened, downloads the
 * ciphertext (which deletes it from the relay), and decrypts it with the
 * link's key and the display fields as additional data. A link whose
 * display fields were altered fails to decrypt.
 */
export async function open(link: string, options: HumanOptions = {}): Promise<OpenResult> {
  const { reveal, pageOrigin } = parseRevealLink(link);
  const relay = options.relay ?? relayOrigin(reveal, pageOrigin);
  const client = makeClient(relay, options);
  let status;
  try {
    status = await client.revealStatus(reveal.id, 0, "", options.signal);
  } catch (err) {
    if (isRelayCode(err, RelayCode.NotFound)) {
      throw new BurndropError("this link has expired or was never created", { cause: err });
    }
    throw err;
  }
  if (status.state !== SlotState.Created) {
    throw new BurndropError("this link was already used or revoked (state: " + status.state + ")");
  }
  const { ciphertext } = await client.open(reveal.id, reveal.revealToken, options.signal);
  let env: Envelope;
  try {
    env = await decryptEnvelope(reveal.key, ciphertext, revealAad(reveal.name, reveal.keepsCopy));
  } catch (err) {
    if (err instanceof RelayError) {
      throw err;
    }
    throw new BurndropError(
      "the secret could not be decrypted: the link was altered or the relay returned the wrong data; the relay copy is gone, ask the agent to send it again",
      { cause: err },
    );
  } finally {
    zero(reveal.key);
  }
  return { name: env.name, value: secretBytes(env), format: env.format, keepsCopy: reveal.keepsCopy, relay };
}

File: sdk/python/src/burndrop/human.py
"""The human side of both flows, for programs that stand in for the page.

:func:`submit` does what the drop page does when the human presses
"Encrypt and send": it seals the value to the key in the link, together
with the metadata the link displayed, and uploads it with the commitment
to that key so an honest relay rejects an altered link. :func:`open` does
what the reveal page does on "Reveal": it fetches the ciphertext (which
deletes it) and decrypts it with the display fields as additional data, so
an altered link fails to decrypt.
"""

from __future__ import annotations

from . import crypto
from .link import normalize_origin, parse_drop, parse_reveal
from .relay import RelayClient

__all__ = ["HumanError", "open", "submit"]


class HumanError(Exception):
    """A reveal could not be decrypted: the link was altered or the relay lied."""


def submit(
    link: str,
    value: bytes,
    *,
    relay: str | None = None,
    client_name: str | None = None,
) -> str:
    """Encrypt ``value`` for the agent behind a drop link and upload it.

    Returns the fingerprint of the key the value was encrypted to; compare
    it with the one the agent showed before calling, as the page asks the
    human to do. ``relay`` overrides the relay origin the link points at.
    """
    drop, page_origin = parse_drop(link)
    origin = normalize_origin(relay) if relay else drop.relay_origin(page_origin)
    value = bytes(value)
    if not value:
        raise ValueError("empty value")
    fingerprint = drop.fingerprint()
    envelope = crypto.Envelope.drop(
        drop.name, drop.purpose, drop.storage, drop.retention, fingerprint, value
    )
    sealed = crypto.seal_envelope(drop.recipient_key, envelope)
    client = RelayClient(origin, client_name=client_name)
    client.upload(drop.id, drop.upload_token, crypto.commitment(drop.recipient_key), sealed)
    return fingerprint


def open(
    link: str,
    *,
    relay: str | None = None,
    client_name: str | None = None,
) -> bytes:
    """Open a reveal link once and return the value.

    The relay deletes the ciphertext when it is handed out, so a second call
    fails with a ``gone`` relay error.
    """
    reveal, page_origin = parse_reveal(link)
    origin = normalize_origin(relay) if relay else reveal.relay_origin(page_origin)
    client = RelayClient(origin, client_name=client_name)
    ciphertext, _created_at = client.open(reveal.id, reveal.reveal_token)
    try:
        envelope = crypto.decrypt_envelope(reveal.key, ciphertext, reveal.aad())
    except crypto.CryptoError as err:
        raise HumanError(
            "the secret could not be decrypted: the link was altered or the relay returned "
            "the wrong data; the relay copy is gone, ask the agent to send it again"
        ) from err
    return envelope.secret_bytes()
